wrangler-core/src/main/java/io/cdap/wrangler/executor/RecipePipelineExecutor.java

```diff
@@ -161,4 +188,55 @@ private List<Directive> getDirectives() throws RecipeException {
     }
     return directives;
   }
+
+  private Schema generateOutputSchema(Schema inputSchema, List<Row> output) throws RecordConvertorException {
+    Map<String, Schema> outputFieldMap = new LinkedHashMap<>();
+    for (Row row : output) {
+      for (Pair<String, Object> rowField : row.getFields()) {
+        String fieldName = rowField.getFirst();
+        Object fieldValue = rowField.getSecond();
+
+        Schema existing = inputSchema.getField(fieldName) != null ? inputSchema.getField(fieldName).getSchema() : null;
+        Schema generated = fieldValue != null && !isValidSchemaForValue(existing, fieldValue) ?
+          generator.getSchema(fieldValue, fieldName) : null;
+
+        if (generated != null) {
+          outputFieldMap.put(fieldName, generated);
+        } else if (existing != null) {
+          outputFieldMap.put(fieldName, existing);
+        }
+      }
+    }
+    List<Field> outputFields = outputFieldMap.entrySet().stream()
+      .map(e -> Schema.Field.of(e.getKey(), e.getValue()))
+      .collect(Collectors.toList());
+    return Schema.recordOf("output", outputFields);
+  }
+
+  // Checks whether the provided input schema is of valid type for given object
+  private boolean isValidSchemaForValue(@Nullable Schema schema, Object value) throws RecordConvertorException {
+    if (schema == null) {
+      return false;
+    }
+    Schema generated = generator.getSchema(value, TEMP_SCHEMA_FIELD_NAME);
+    generated = generated.isNullable() ? generated.getNonNullable() : generated;
+    schema = schema.isNullable() ? schema.getNonNullable() : schema;
+    return generated.getType().equals(schema.getType());
+  }
+
+  // Gets the union of fields in two schemas while maintaining insertion order and uniqueness of fields. If the same
+  // field exists with two different schemas, the second schema overwrites first one
+  private Schema getSchemaUnion(Schema first, Schema second) {
```

Comment:
If which cases would it be useful? I believe we always generate schema for all field of the row. 
E.g. if original schema of the field is "string" and resulting generated schema is "nullable number", this algorythm with randomly (depending on a sample) generate either "string" or "nullable number"

Reply:
That's how it is currently done (without schema addition). The last row of the sample's data is ultimately used to determine the type. So in this case, we do a schema union for a different reason. Lets say we have the json parsing directive which generates two rows with these columns:
- col_A: int, col_B: string
- col_B: string, col_C: float

Now, after the schema has been taken at the end of executing the directives on the first row, it will have columns A and B. We need to take a union after executing the directives on the second row as well to pick up all 3 columns in the schema.